File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/ExternalOAuthServiceImpl.java
package com.skedgo.tripkit.booking;

import android.text.TextUtils;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;

public class ExternalOAuthServiceImpl implements ExternalOAuthService {
    private final ExternalOAuthServiceGenerator externalOAuthServiceGenerator;

    public ExternalOAuthServiceImpl(ExternalOAuthServiceGenerator externalOAuthServiceGenerator) {
        this.externalOAuthServiceGenerator = externalOAuthServiceGenerator;
    }

    @Override
    public Observable<ExternalOAuth> getAccessToken(
        final BookingForm form,
        final String code,
        final String grantType,
        final String callback
    ) {
        final String clientId = form.getClientID();
        final String clientSecret = form.getClientSecret();
        final String baseUrl = form.getTokenURL();
        final String scope = TextUtils.isEmpty(form.getScope()) ? null : form.getScope();
        final Object serviceId = form.getValue();

        // TODO: how to avoid this constant?
        boolean credentialsInHeader = serviceId == null || !("uber".equals(serviceId.toString()));

        final ExternalOAuthApi externalOAuthApi =
            externalOAuthServiceGenerator.createService(baseUrl, clientId,
                clientSecret, credentialsInHeader
            );

        return externalOAuthApi.getAccessToken(clientSecret, clientId, code, "authorization_code", callback, scope)
            .filter(new Predicate<AccessTokenResponse>() {
                @Override
                public boolean test(AccessTokenResponse response) {
                    return response != null && response.refreshToken() != null;
                }
            })
            .map(new Function<AccessTokenResponse, ExternalOAuth>() {
                @Override
                public ExternalOAuth apply(AccessTokenResponse response) {
                    String serviceIdString = "";
                    if (serviceId != null) {
                        serviceIdString = serviceId.toString();
                    }
                    return ImmutableExternalOAuth.builder()
                        .authServiceId(serviceIdString)
                        .token(response.accessToken())
                        .refreshToken(response.refreshToken())
                        .expiresIn(response.expiresIn())
                        .build();
                }
            });
    }
}

File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/ExternalOAuthServiceImpl.kt
package com.skedgo.tripkit.booking

import io.reactivex.Observable

class ExternalOAuthServiceImpl(
    private val externalOAuthServiceGenerator: ExternalOAuthServiceGenerator
) : ExternalOAuthService {

    override fun getAccessToken(
        form: BookingForm,
        code: String,
        grantType: String,
        callback: String
    ): Observable<ExternalOAuth> {
        val clientId = form.getClientID()
        val clientSecret = form.getClientSecret()
        val baseUrl = form.getTokenURL()
        val scope = if (form.getScope().isNullOrEmpty()) null else form.getScope()
        val serviceId = form.mValue

        // TODO: how to avoid this constant?
        val credentialsInHeader = serviceId == null || serviceId.toString() != "uber"

        val externalOAuthApi = externalOAuthServiceGenerator.createService(
            baseUrl.orEmpty(),
            clientId,
            clientSecret,
            credentialsInHeader
        )

        return externalOAuthApi.getAccessToken(
            clientSecret,
            clientId,
            code,
            "authorization_code",
            callback,
            scope
        )
            .filter { response -> response != null && response.refreshToken() != null }
            .map { response ->
                val serviceIdString = serviceId?.toString() ?: ""
                ImmutableExternalOAuth.builder()
                    .authServiceId(serviceIdString)
                    .token(response.accessToken())
                    .refreshToken(response.refreshToken())
                    .expiresIn(response.expiresIn())
                    .build()
            }
    }
}